README.md
# io_import_rbsp
Blender 2.93 importer for `rBSP` files (Titanfall Engine `.bsp`)


<!-- TODO: .gifs for guides -->
## Installation
 * Get Blender 2.93+ from [Steam](https://store.steampowered.com/app/365670/Blender/) / [Blender.org](https://www.blender.org/download/)
 * Head over to [releases](https://github.com/snake-biscuits/io_import_rbsp/releases/) & download `io_import_rbsp_v1.2.0_b2.93.zip`  
Then, in Blender:  
 * `Edit > Preferences > Add-ons > Install`
 * Find `io_import_rbsp_v1.2.0_b2.93.zip`
 * Click `Install Addon`
 * Check the box to enable `Import-Export: io_import_rbsp`


## Usage

> WARNING: Titanfall Engine maps are huge  
> Imports can take multiple minutes and a few GB of RAM for the geometry alone  
> Test a small map before loading Olympus and setting your PC on fire

### Extracting `.bsp`s
You will need to extract `.bsp`, `.bsp_lump` & `.ent` files for any map you want to extract. To do this:
 * Grab a [Respawn VPK extractor](#Respawn-VPK)
 * Locate the `.vpk`s for the game you want to work with (game must be installed)
   - `Titanfall/vpk/`
   - `Titanfall2/vpk/`
   - `Apex Legends/vpk/`
 * Open the `*.bsp.pak000_dir.vpk` for the map you want to load
   - Titanfall 2 map names can be found here: [NoSkill Modding Wiki](https://noskill.gitbook.io/titanfall2/documentation/file-location/vpk-file-names)
   - Lobbies are stored in `mp_common.bsp.pak000_dir.vpk`
 * Extract the `.bsp`, `.ent`s & `.bsp_lumps` from the `maps/` folder to someplace you'll remember
   - each `.vpk` holds assets for one `.bsp` (textures and models are stored elsewhere)
<!-- TODO: Materials & Models -->

### Blender Importer
Once you've extracted the files you need:
 * `File > Import > Titanfall Engine .bsp`
 * Select the `.bsp` (`.bsp_lump` & `.ent` files need to be in the same folder)
 * Choose your settings
 * Click Import
 * Wait a few minutes (Can easily take 1hr+ on Apex Legends maps)
<!-- Cry when it breaks and send me an e-mail to fix it -->
<!-- Materials & Models will require paths set in Preferences & SourceIO -->

### Legion Assets
**TITANFALL 2 & APEX LEGENDS ONLY! (EXPERIMENTAL)**
- Titanfall: `.vpk` Materials & Models
- Titanfall 2: `.rpak` Materials; `.vpk` Models
- Apex: `.rpak` Materials & Models
 * [Legion Workflow](https://drive.google.com/file/d/1ApByE0p5MzVV95dUsQ0seciCA7Cl5WFZ/view)
 * Requires:
   - [.cast](https://github.com/dtzxporter/cast)
   - [.mprt](https://github.com/llennoco22/Apex-mprt-importer-for-Blender)



## Related Tools

### Respawn VPK
 * [Titanfall_VPKTool3.4_Portable](https://github.com/Wanty5883/Titanfall2/blob/master/tools/Titanfall_VPKTool3.4_Portable.zip) (GUI only)
   - by `Cra0kalo` (currently Closed Source) **recommended**
 * [TitanfallVPKTool](https://github.com/p0358/TitanfallVPKTool) (GUI & CLI)
   - by `P0358`
 * [RSPNVPK](https://github.com/squidgyberries/RSPNVPK) (CLI only)
   - Fork of `MrSteyk`'s Tool
 * [UnoVPKTool](https://github.com/Unordinal/UnoVPKTool) (CLI only)
   - by `Unordinal`

### Other
 * [SourceIO](https://github.com/REDxEYE/SourceIO)
   - GoldSrc & Source Engine importer (`.bsp`, `.vmt`, `.vtf`, `.mdl`)
 * [SourceOps](https://github.com/bonjorno7/SourceOps)
   - Source Engine model exporter
 * [PyD3DBSP](https://github.com/mauserzjeh/PyD3DBSP) (Archived)
   - Call of Duty 2 `.bsp` importer
 * [blender_io_mesh_bsp](https://github.com/andyp123/blender_io_mesh_bsp)
   - Quake 1 `.bsp` importer
 * [Blender_BSP_Importer](https://github.com/QuakeTools/Blender_BSP_Importer)
   - Quake 3 `.bsp` importer


## FAQs
 * Why can't I see anything?
   - Titanfall Engine maps are huge, you need to increase your view distance
   - `3D View > N > View > Clip Start: 16, End: 102400` (only affects that 3D view)
     - You will also need to increase the clipping distance for all cameras
 * Why is my `.blend` file still huge after I deleted everything?
   - Blender keeps deleted items cached in case you want to undo
   - To clear this cache use: `File > Clean Up > Recursive Unused Data Blocks`
   - Or set the **Outliner** display mode to **Orphan Data** & click **Purge**
 * It broke? Help?
   - Ask around on Discord, you might've missed a step someplace
   - If you're loading a brand new Apex map, it might not be supported yet
 * Can I use this to make custom maps?
   - No, we don't know enough about Respawn's `.bsp` format to make compilers
   - As easy as it might sound on paper, editing a `.bsp` directly is no small task
 * Can I use this for animations?
   - Sure! but be sure to credit the tool someplace
   - And credit Respawn too! they made the maps in the first place

### Further Questions
I can be found on the following Titanfall related Discord Servers as `b!scuit#3659`:
  * Titanfall 1:
    - [TF Remnant Fleet](https://discord.gg/hKpQeJqdZR)
  * Titanfall 2:
    - Titanfall 2 Speedrunning Discord
  * Apex Legends:
    - [R5Reloaded](https://discord.com/invite/jqMkUdXrBr)
<!-- TODO: add Titanfall Online & Titanfall 2 Custom Servers when they go public -->

> NOTE: I am a fully time Uni Student in an Australian Timezone  
> Don't go expecting an immediate response


CHANGELOG.md
# Changelog

## v1.2.2_b2.93 (8th December 2021)

### Updates
 * Models & Textures via [Legion](https://wiki.modme.co/wiki/apps/Legion.html) **(experimental)**; requires:
   - [.cast](https://github.com/dtzxporter/cast)
   - [.mprt](https://github.com/llennoco22/Apex-mprt-importer-for-Blender)

> Faster methods (as part of io_import_rbsp) are planned, this was just easier to implement


## v1.2.1_b2.93 (8th December 2021)

### Updates
 * Updated internal `bsp_tool` to [latest version](https://github.com/snake-biscuits/bsp_tool/commit/13836462855b4cbd8049098a1df350b71eb1094f)

### Fixed
 * Apex Legends Season 11 (post 19th November path) maps are now correctly detected


## v1.2.0_b2.93 (1st October 2021)

### Added
 * Entity naming resolution order is now `targetname -> editorclass -> classname`
 * All brush entities now generate geo and are coloured from a hard-coded palette

### Fixed
 * Generated trigger brush geo is no longer inside-out


## v1.1.0_b2.93 (1st October 2021)

### Added
 * Geometry for `trigger_*` brush entities is now generated on import


## v1.0.0_b2.93 (28th September 2021)
Initial Release


io_import_rbsp/__init__.py
import bpy
from bpy_extras.io_utils import ImportHelper
from bpy.props import StringProperty, BoolProperty, EnumProperty
from bpy.types import Operator

from . import bsp_tool
from . import rbsp
from . import preferences


bl_info = {
    "name": "io_import_rbsp",
    "author": "Jared Ketterer (snake-biscuits / Bikkie)",
    "version": (1, 3, 0),
    "blender": (2, 93, 6),
    "location": "File > Import > Titanfall Engine .bsp",
    "description": "Import maps from Titanfall, Titanfall 2 & Apex Legends",
    "doc_url": "https://github.com/snake-biscuits/io_import_rbsp",
    "tracker_url": "https://github.com/snake-biscuits/io_import_rbsp/issues",
    "category": "Import-Export"}


class ImportRBSP(Operator, ImportHelper):
    """Load Titanfall Engine rBSP"""
    bl_idname = "io_import_rbsp.rbsp_import"
    bl_label = "Titanfall Engine .bsp"
    filename_ext = ".bsp"
    filter_glob: StringProperty(default="*.bsp", options={"HIDDEN"}, maxlen=255)  # noqa F722
    # TODO: load_materials EnumProperty: None, Names, Base Colours, Nodes
    load_geometry: BoolProperty(name="Geometry", description="Load .bsp Geometry", default=True)  # noqa F722
    # TODO: load_lighting EnumProperty: None, Empties, All, PortalLights
    load_entities: BoolProperty(name="Entities", description="Load .bsp Entities", default=True)  # noqa F722
    load_brushes: EnumProperty(name="Brushes", description="Generate brush geometry",  # noqa F722
                               items=(("None", "No Brushes", "Empties at brush entity origins"),  # noqa F722
                                      ("Brushes", "Generate Brushes", "Generate brushes from planes stored in the entity")),  # noqa F722
                               default="Brushes")  # noqa F722
    # TODO: load_prop_dynamic EnumProperty: None, Empties, Low-Poly, High-Poly
    # TODO: load_prop_static EnumProperty: None, Empties, Low-Poly, High-Poly, Skybox Only
    load_props: EnumProperty(name="Props", description="Load smaller models",  # noqa F722
                             items=(("None", "No Props", "Decent performance"),  # noqa F722
                                    ("Empties", "Use Empties", "Place empties at prop origins"),  # noqa F722
                                    ("Instances", "Instanced Collections", "Full props & materials; Very slow")),  # noqa F722
                             default="None")  # noqa F722
    # TODO: cubemap volumes?
    # TODO: Lock out some options if SourceIO is not installed, alert the user
    # TODO: Warnings for missing files
    # TODO: Lightmaps (export to file(s) and append to materials (would require splitting materials up more))

    def execute(self, context):
        """Import selected .bsp"""
        # TODO: loading % indicators
        bsp = bsp_tool.load_bsp(self.filepath)
        import_script = {bsp_tool.branches.respawn.titanfall: rbsp.titanfall,
                         bsp_tool.branches.respawn.titanfall2: rbsp.titanfall2,
                         bsp_tool.branches.respawn.apex_legends: rbsp.apex_legends}
        importer = import_script[bsp.branch]
        # master_collection
        if bsp.filename not in bpy.data.collections:
            master_collection = bpy.data.collections.new(bsp.filename)
            bpy.context.scene.collection.children.link(master_collection)
        else:
            master_collection = bpy.data.collections[bsp.filename]
        # materials
        materials = importer.materials.base_colours(bsp)
        # if bsp.branch == bsp_tool.branches.respawn.titanfall2:  # HACK
        #     materials = importer.materials.textured(bsp)
        # geometry
        if self.load_geometry:
            # TODO: rename Model[0] "worldspawn"
            # TODO: skybox collection
            # TODO: load specific model / mesh (e.g. worldspawn only, skip tool brushes etc.)
            importer.geometry(bsp, master_collection, materials)
        # entities
        # TODO: get a user defined palette in here (use a property group?)
        # -- FloatVectorProperty(subtype="COLOR")  # ?
        brush_entity_classnames = ["envmap_volume",
                                   "light_probe_volume",
                                   "light_environment_volume",
                                   "trigger_capture_point",
                                   "trigger_hurt",
                                   "trigger_indoor_area",
                                   "trigger_multiple",
                                   "trigger_once",
                                   "trigger_out_of_bounds",
                                   "trigger_soundscape"]
        if self.load_brushes == "Brushes":
            brushes_entities = {classname: importer.entities.trigger_brushes for classname in brush_entity_classnames}
            importer.entities.ent_object_data.update(brushes_entities)
        if self.load_entities:
            # TODO: link worldspawn to Model[0]
            importer.entities.all_entities(bsp, master_collection)
            # NOTE: includes lights, square info_node* & speakers
            # NOTE: Eevee has limited lighting, try Cycles
        # props
        if self.load_props == "Empties":
            importer.props.as_empties(bsp, master_collection)
        elif self.load_props == "Instances":
            importer.props.static_props(bsp, master_collection)
        # TODO: scale the whole import (Engine Units -> Inches)
        # TODO: override default view / camera (near, far) clipping -> (16, 102400)
        return {"FINISHED"}


# Only needed if you want to add into a dynamic menu
def menu_func_import(self, context):
    self.layout.operator(ImportRBSP.bl_idname, text=ImportRBSP.bl_label)


def register():
    preferences.register()
    bpy.utils.register_submodule_factory(__name__, ("bsp_tool", "rbsp"))
    bpy.utils.register_class(ImportRBSP)
    bpy.types.TOPBAR_MT_file_import.append(menu_func_import)


def unregister():
    preferences.unregister()
    bpy.utils.unregister_class(ImportRBSP)
    bpy.types.TOPBAR_MT_file_import.remove(menu_func_import)


if __name__ == "__main__":
    register()
